### PR TITLE
implement LineMarkers with implementation data

### PR DESCRIPTION
**Pull request by swissiety**
addresses #238.

### CODE REVIEW

File: src/main/java/org/wso2/lsp4intellij/contributors/psi/LSPPsiSymbol.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+import com.intellij.icons.AllIcons;
+import com.intellij.navigation.ItemPresentation;
+import com.intellij.openapi.project.Project;
+import com.intellij.psi.PsiFile;
+import org.eclipse.lsp4j.SymbolKind;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
```
Comment (nixel2007) [inline src/main/java/org/wso2/lsp4intellij/contributors/psi/LSPPsiSymbol.java:26]: swing? O_o

Comment (swissiety) [reply src/main/java/org/wso2/lsp4intellij/contributors/psi/LSPPsiSymbol.java]: for Icon

File: src/main/java/org/wso2/lsp4intellij/contributors/psi/LSPPsiSymbol.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import com.intellij.navigation.ItemPresentation;
+import com.intellij.openapi.project.Project;
+import com.intellij.psi.PsiFile;
+import org.eclipse.lsp4j.SymbolKind;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.Icon;
+
+public class LSPPsiSymbol extends LSPPsiElement{
```
Comment (NipunaRanasinghe) [inline src/main/java/org/wso2/lsp4intellij/contributors/psi/LSPPsiSymbol.java:28]:
```suggestion
public class LSPPsiSymbol extends LSPPsiElement {
```

File: src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.editor.EditorFactory;
+import com.intellij.openapi.fileEditor.FileDocumentManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import org.eclipse.lsp4j.*;
```
Comment (NipunaRanasinghe) [inline src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java:35]: Lets avoid wildcard imports.

File: src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+import org.wso2.lsp4intellij.client.languageserver.requestmanager.RequestManager;
+import org.wso2.lsp4intellij.client.languageserver.wrapper.LanguageServerWrapper;
+import org.wso2.lsp4intellij.contributors.psi.LSPPsiElement;
+import org.wso2.lsp4intellij.contributors.psi.LSPPsiSymbol;
+import org.wso2.lsp4intellij.requests.Timeout;
+import org.wso2.lsp4intellij.requests.Timeouts;
+import org.wso2.lsp4intellij.utils.DocumentUtils;
+import org.wso2.lsp4intellij.utils.FileUtils;
+
+import java.util.*;
```
Comment (NipunaRanasinghe) [inline src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java:48]: Lets avoid wildcard imports.

File: src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        startPos = docSymbol.getRange().getStart();
+        final int start = DocumentUtils.LSPPosToOffset(editor, startPos);
+        final int end = DocumentUtils.LSPPosToOffset(editor, docSymbol.getRange().getEnd());
+        if (start < 0 || end < 0) {
+          continue;
+        }
+        lspPsiSymbol = new LSPPsiSymbol(docSymbol.getKind(), docSymbol.getName(), project, start, end, containingFile);
+        augmentSymbol(result, wrapper, textDocument, requestManager, project, lspPsiSymbol, startPos);
+      }
+
```
Comment (NipunaRanasinghe) [inline src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java:162]: Redundant empty line

File: src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
```
Comment (NipunaRanasinghe) [inline src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java:164]: Redundant empty line

File: src/main/java/org/wso2/lsp4intellij/contributors/symbol/WorkspaceSymbolProvider.java [hunk 7]
```diff
@@ -83,24 +84,26 @@ private LSPNavigationItem createNavigationItem(LSPSymbolResult result, Project p
...
     }
 
     try {
       List<? extends SymbolInformation> symbolInformations = request
-          .get(20000, TimeUnit.MILLISECONDS);
-      wrapper.notifySuccess(Timeouts.SYMBOLS);
-      return symbolInformations.stream()
-          .map(si -> new LSPSymbolResult(si, wrapper.getServerDefinition()));
+              .get(Timeout.getTimeout(Timeouts.WORKSPACESYMBOLS), TimeUnit.MILLISECONDS);
+      wrapper.notifySuccess(Timeouts.WORKSPACESYMBOLS);
```
Comment (NipunaRanasinghe) [inline src/main/java/org/wso2/lsp4intellij/contributors/symbol/WorkspaceSymbolProvider.java:99]:
```suggestion
      wrapper.notifySuccess(Timeouts.WORKSPACE_SYMBOLS);
```

File: src/main/java/org/wso2/lsp4intellij/contributors/symbol/WorkspaceSymbolProvider.java [hunk 8]
```diff
@@ -83,24 +84,26 @@ private LSPNavigationItem createNavigationItem(LSPSymbolResult result, Project p
...
+      if( symbolInformations != null){
+        return symbolInformations.stream()
+                .map(si -> new LSPSymbolResult(si, wrapper.getServerDefinition()));
+      }
     } catch (TimeoutException e) {
       LOG.warn(e);
-      wrapper.notifyFailure(Timeouts.SYMBOLS);
+      wrapper.notifyFailure(Timeouts.WORKSPACESYMBOLS);
```
Comment (NipunaRanasinghe) [inline src/main/java/org/wso2/lsp4intellij/contributors/symbol/WorkspaceSymbolProvider.java:106]:
```suggestion
      wrapper.notifyFailure(Timeouts.WORKSPACE_SYMBOLS);
```

File: src/main/java/org/wso2/lsp4intellij/requests/Timeouts.java [hunk 9]
```diff
@@ -21,7 +21,7 @@
 public enum Timeouts {
     CODEACTION(2000), CODELENS(2000), COMPLETION(1000), DEFINITION(2000), DOC_HIGHLIGHT(1000), EXECUTE_COMMAND(
             2000), FORMATTING(2000), HOVER(2000), INIT(10000), REFERENCES(2000), SIGNATURE(1000), SHUTDOWN(
-            5000), SYMBOLS(2000), WILLSAVE(2000);
+            5000), SYMBOLS(2000), WORKSPACESYMBOLS(5000), WILLSAVE(2000), IMPLEMENTATION(2000);
```
Comment (NipunaRanasinghe) [inline src/main/java/org/wso2/lsp4intellij/requests/Timeouts.java:24]:
```suggestion
            5000), SYMBOLS(2000), WORKSPACE_SYMBOLS(5000), WILLSAVE(2000), IMPLEMENTATION(2000);
```

**Review by NipunaRanasinghe — COMMENTED** [review]
@swissiety Thanks for your time working on this and the PR looks good, except for the minor suggestions I have added.

**Comment by NipunaRanasinghe** [comment]
@swissiety Will you have some spare time to address the review suggestions and make this ready for review? :slightly_smiling_face:

File: src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        if (symbolInfo.getKind() != SymbolKind.Method && symbolInfo.getKind() != SymbolKind.Constructor && symbolInfo.getKind() != SymbolKind.Class && symbolInfo.getKind() != SymbolKind.Interface) {
+          continue;
+        }
+        startPos = symbolInfo.getLocation().getRange().getStart();
+        final int start = DocumentUtils.LSPPosToOffset(editor, startPos);
+        final int end = DocumentUtils.LSPPosToOffset(editor, symbolInfo.getLocation().getRange().getEnd());
+        if (start < 0 || end < 0) {
+          continue;
+        }
+        lspPsiSymbol = new LSPPsiSymbol(symbolInfo.getKind(), symbolInfo.getName(), project, start, end, containingFile);
```
Comment (Trias) [inline src/main/java/org/wso2/lsp4intellij/contributors/symbol/LineMarkerProvider.java:146]: I'm working on a similar FEature and was using your PR as a reference, Thank you :)

I solved this particular line differently (Kotlin code, but i think it's)
```
File containingFile = elements[0].containingFile;
PsiSymbol symbol = containingFile.findElementAt(start);

if (symbol == null || !elements.contains(symbol)) {
    continue
}
```

this way we don't need the check in line 79 (which was not working for me..) and intellij reacts quickly on user input

**Pull request closed without merging**